Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// Support functions for liburing
+
+// Check if this is needed, copied from shim.c to avoid possible problems due to:
+// Xcode's Archive builds with Xcode's Package support struggle with empty .c files
+// (https://bugs.swift.org/browse/SR-12939).
+void CNIOLinux_i_do_nothing_just_working_around_a_darwin_toolchain_bug2(void) {}
+
+#ifdef __linux__
+
+#ifdef SWIFTNIO_LIBURING
+
+#define _GNU_SOURCE
+#include <CNIOLinux.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <signal.h>
+#include <signal.h>
+#include <sys/poll.h>
+
+// Definitons from syscall.c in liburing, we'll just do a syscall that will fail
+// to check for availability of the interface - better than checking specific kernel
+// versions as support can be backported to earlier kernels
+
+#ifdef __alpha__
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    537
+# endif
+#else /* !__alpha__ */
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    427
+# endif
+#endif
+
+int _check_syscall_available() {
+    errno = 0;
+    if (syscall(__NR_io_uring_register, 0, IORING_UNREGISTER_BUFFERS, NULL, 0) && errno == ENOSYS) {
+        return -1;
+    }
+    return 0;
+}
+
+int _check_capabilities() {
+    struct io_uring_probe *probe = io_uring_get_probe();
+    int capabilities_check;
+
+    capabilities_check = io_uring_opcode_supported(probe, IORING_OP_POLL_ADD) &
+                         io_uring_opcode_supported(probe, IORING_OP_POLL_REMOVE);
+
+    io_uring_free_probe(probe);
+    return capabilities_check == 1 ? 0 : -1;
+}
+
+int CNIOLinux_check_compatibility()
+{
+    if (_check_syscall_available() != 0) {
+        return -1;
+    }
+    if (_check_capabilities() != 0) {
+        return -1;
+    }
+    return 0;
+}
+
+// We should look at cpu affinity and preferably bind the SQPOLL thread to a specific core
+// and support cgroup / cpu sets on Linux to isolate it - then we should set IORING_SETUP_SQ_AFF here.
+
+static pthread_mutex_t global_ring_mutex = PTHREAD_MUTEX_INITIALIZER;
+static struct io_uring global_ring; // shared small ring to be able to reuse SQPOLL kernel thread
+int CNIOLinux_io_uring_queue_init(unsigned entries, struct io_uring *ring,
+    unsigned flags)
+{
+    if (flags & IORING_SETUP_SQPOLL)
```

Comment:
what's the current state w.r.t. SQPOLL requiring `root`?

Reply:
It is [planned to be lifted in 5.13](https://git.kernel.dk/cgit/linux-block/commit/?h=for-5.13/io_uring&id=548d819d1eed7b6bf86d36c8de2fbc54b69db571) which is a little bit out, 5.12 is not shipped yet...

Reply:
@Lukasa WDYT, should we still ship SQPOLL support despite straight away the fact that even with the latest-greatest it needs `root`? Don't really want to encourage people to run NIO programs as root. But of course we can just take the code as is and runtime detect. But I'd assume nobody to actually be running `SQPOLL` (because `root`) so we've more code that's less frequently run so more permutations?

Reply:
In current kernels, if non-privileged, it will just work as usual and disregard the sqpoll flag IIRC.

Reply:
Yes, that's my understanding too. What I was trying to say is given that (hopefully) nobody runs this as root, nobody will exercise the `SQPOLL` code. So maybe we should have the initial version without `SQPOLL` support at all and at that later (once kernels support it).

Reply:
I'm inclined to drop it and add it back later once we think folks can actually exercise the code.

Reply:
Ok,, code removed.